selene-lib/src/lints/roblox_suspicious_udim2_new.rs

```diff
@@ -0,0 +1,138 @@
+        if !context.is_roblox() {
+            return Vec::new();
+        }
+
+        let mut visitor = UDim2CountVisitor::default();
+
+        visitor.visit_ast(ast);
+
+        visitor
+            .args
+            .iter()
+            .map(|mismatch| {
+                Diagnostic::new_complete(
+                    "roblox_suspicious_udim2_new",
+                    format!(
+                        "UDim2.new takes 4 numbers, but {} {} provided.",
+                        mismatch.args_provided,
+                        if mismatch.args_provided == 1 {
+                            "was"
+                        } else {
+                            "were"
+                        }
+                    ),
+                    Label::new(mismatch.call_range),
+                    vec![
+                        if mismatch.args_provided > 2 || !mismatch.args_are_numbers {
+                            ""
+                        } else if mismatch.args_are_between_0_and_1 {
+                            "did you mean to use UDim2.fromScale instead?"
+                        } else {
+                            "did you mean to use UDim2.fromOffset instead?"
+                        }
+                        .to_owned(),
+                    ],
+                    Vec::new(),
+                )
+            })
+            .collect()
+    }
+}
+
+#[derive(Default)]
+struct UDim2CountVisitor {
+    args: Vec<MismatchedArgCount>,
+}
+
+struct MismatchedArgCount {
+    args_provided: usize,
+    call_range: (usize, usize),
+    args_are_between_0_and_1: bool,
+    args_are_numbers: bool,
+}
+
+impl Visitor for UDim2CountVisitor {
+    fn visit_function_call(&mut self, call: &ast::FunctionCall) {
+        if_chain::if_chain! {
+            if let ast::Prefix::Name(token) = call.prefix();
+            if token.token().to_string() == "UDim2";
+            let mut suffixes = call.suffixes().collect::<Vec<_>>();
+
+            if suffixes.len() == 2; // .new and ()
+            let call_suffix = suffixes.pop().unwrap();
+            let index_suffix = suffixes.pop().unwrap();
+
+            if let ast::Suffix::Index(ast::Index::Dot { name, .. }) = index_suffix;
+            if name.token().to_string() == "new";
+
+            if let ast::Suffix::Call(ast::Call::AnonymousCall(
+                ast::FunctionArgs::Parentheses { arguments, .. }
+            )) = call_suffix;
+
+            then {
+                let args_provided = arguments.len();
+                let numbers_passed = arguments.iter().filter(|expression| {
+                    match expression {
+                        ast::Expression::Value { value, .. } => matches!(&**value, ast::Value::Number(_)),
+                        _ => false,
+                    }
+                }).count();
+
+                if args_provided == 2 && numbers_passed == 0 {
+                    return;
+                };
+
+                if args_provided < 4 {
```

Comment:
Before performing the semi-expensive O(n) work, let's early return if `args_provided == 4`.